File: backend/src/main/java/com/project/controller/MoodController.java
package com.project.controller;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.project.dto.mood.request.MoodRequestDTO;
import com.project.dto.mood.response.MoodAnalyticsResponseDTO;
import com.project.dto.mood.response.MoodResponseDTO;
import com.project.exception.ApiResponse;
import com.project.exception.ResponseBuilder;
import com.project.security.CustomUserDetails;
import com.project.service.MoodService;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;

@RestController
@RequestMapping("/api/moods")
@RequiredArgsConstructor
public class MoodController {

    private final MoodService moodService;

    /* ================= ADD MOOD ================= */

    @PostMapping
    public ResponseEntity<ApiResponse<MoodResponseDTO>> addMood(
            @RequestBody @Valid MoodRequestDTO dto,
            Authentication auth) {

        CustomUserDetails user =
                (CustomUserDetails) auth.getPrincipal();

        MoodResponseDTO response =
                moodService.addTodayMood(dto, user.getUserId());

        return ResponseBuilder.success(
                "Mood added successfully",
                response,
                HttpStatus.CREATED
        );
    }

    /* ================= UPDATE MOOD ================= */

    @PutMapping
    public ResponseEntity<ApiResponse<MoodResponseDTO>> updateMood(
            @RequestBody @Valid MoodRequestDTO dto,
            Authentication auth) {

        CustomUserDetails user =
                (CustomUserDetails) auth.getPrincipal();

        MoodResponseDTO response =
                moodService.updateTodayMood(dto, user.getUserId());

        return ResponseBuilder.success(
                "Mood updated successfully",
                response,
                HttpStatus.OK
        );
    }

    /* ================= GET MY MOODS ================= */

    @GetMapping
    public ResponseEntity<ApiResponse<List<MoodResponseDTO>>> getMyMoods(
            Authentication auth) {

        CustomUserDetails user =
                (CustomUserDetails) auth.getPrincipal();

        List<MoodResponseDTO> moods =
                moodService.getUserMoods(user.getUserId());

        return ResponseBuilder.success(
                "User moods fetched successfully",
                moods,
                HttpStatus.OK
        );
    }

    /* ================= WEEKLY ANALYTICS ================= */

    @GetMapping("/weekly")
    public ResponseEntity<ApiResponse<MoodAnalyticsResponseDTO>> weekly(
            Authentication auth) {

        CustomUserDetails user =
                (CustomUserDetails) auth.getPrincipal();

        MoodAnalyticsResponseDTO response =
                moodService.getWeeklyAnalytics(user.getUserId());

        return ResponseBuilder.success(
                "Weekly mood analytics fetched successfully",
                response,
                HttpStatus.OK
        );
    }

    /* ================= MONTHLY ANALYTICS ================= */

    @GetMapping("/monthly")
    public ResponseEntity<ApiResponse<MoodAnalyticsResponseDTO>> monthly(
            @RequestParam int year,
            @RequestParam int month,
            Authentication auth) {

        CustomUserDetails user =
                (CustomUserDetails) auth.getPrincipal();

        MoodAnalyticsResponseDTO response =
                moodService.getMonthlyAnalytics(
                        user.getUserId(), year, month);

        return ResponseBuilder.success(
                "Monthly mood analytics fetched successfully",
                response,
                HttpStatus.OK
        );
    }
}

File: backend_dotnet/SafeMind.API/Controllers/MoodController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using SafeMind.Core.DTOs.Mood.Request;
using SafeMind.Core.Interfaces;
using System.Security.Claims;
using SafeMind.Core.Responses;

namespace SafeMind.API.Controllers
{
    [ApiController]
    [Route("api/moods")]
    [Authorize]
    public class MoodController : ControllerBase
    {
        private readonly IMoodService _moodService;

        public MoodController(IMoodService moodService)
        {
            _moodService = moodService;
        }

        private long GetCurrentUserId()
        {
            var userIdClaim = User.FindFirst("userId");
            if (userIdClaim == null) throw new UnauthorizedAccessException("Invalid token");
            return long.Parse(userIdClaim.Value);
        }

        [HttpPost]
        public async Task<IActionResult> AddMood([FromBody] MoodRequestDTO dto)
        {
             try
            {
                var userId = GetCurrentUserId();
                var response = await _moodService.AddTodayMoodAsync(dto, userId);
                
                return CreatedAtAction(nameof(GetMyMoods), new { }, 
                    ApiResponse<SafeMind.Core.DTOs.Mood.Response.MoodResponseDTO>.SuccessResponse("Mood added successfully", response));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpPut]
        public async Task<IActionResult> UpdateMood([FromBody] MoodRequestDTO dto)
        {
             try
            {
                var userId = GetCurrentUserId();
                var response = await _moodService.UpdateTodayMoodAsync(dto, userId);
                
                return Ok(ApiResponse<SafeMind.Core.DTOs.Mood.Response.MoodResponseDTO>.SuccessResponse("Mood updated successfully", response));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpGet]
        public async Task<IActionResult> GetMyMoods()
        {
             try
            {
                var userId = GetCurrentUserId();
                var moods = await _moodService.GetUserMoodsAsync(userId);
                
                return Ok(ApiResponse<List<SafeMind.Core.DTOs.Mood.Response.MoodResponseDTO>>.SuccessResponse("User moods fetched successfully", moods));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpGet("weekly")]
        public async Task<IActionResult> Weekly()
        {
             try
            {
                var userId = GetCurrentUserId();
                var response = await _moodService.GetWeeklyAnalyticsAsync(userId);
                
                return Ok(ApiResponse<SafeMind.Core.DTOs.Mood.Response.MoodAnalyticsResponseDTO>.SuccessResponse("Weekly mood analytics fetched successfully", response));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpGet("monthly")]
        public async Task<IActionResult> Monthly([FromQuery] int year, [FromQuery] int month)
        {
             try
            {
                var userId = GetCurrentUserId();
                var response = await _moodService.GetMonthlyAnalyticsAsync(userId, year, month);
                
                return Ok(ApiResponse<SafeMind.Core.DTOs.Mood.Response.MoodAnalyticsResponseDTO>.SuccessResponse("Monthly mood analytics fetched successfully", response));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }
    }
}
